[COT] Implement create() method in COT datastore

jorgeatorres opened the pull request:
### Description
This PR implements the `create()` method in the COT datastore, allowing new orders to be persisted to the database by calling `$order->save()`.

It enhances the functionality first introduced in #33026.

Some additions in this PR:

- Reworked some of the logic that persists data into the database to simplify it.
- Overrides and makes use of `update_post_meta()` to sync metadata with some COT fields.
   This also improves backwards compatibility because metadata can be accessed and directly updated for fields like `order_stock_reduced` and it'd get persisted to the correct table.
- Added unit test `test_cot_datastore_create` which creates an order from scratch, reads it fresh from the database and compares values with expected ones.

#### ⚠️ For discussion
- This PR still treats some props (such as `order_stock_reduced`) as a special case, as those haven't been turned into `WC_Order` properties yet. This works fine, but it's worth mentioning, this is still in the COT blockers list for discussion.
- The `create()` method in this datastore behaves as in current data stores in the sense that it  determines whether a record is new or not based on whether it has an ID or not.
   The current CPT datastore, for example, doesn't allow you to insert a record with a pre-defined ID, but we might need this for the COT datastore in order to reuse IDs from the `wp_posts` table, as long as that's necessary.
   I mostly focused on getting the basics working and feature parity with the CPT datastore, but I can work on that functionality if we need at this point (maybe in a different PR?). It wouldn't be too difficult to add it IMO.

Closes #32670.

### Testing
1. Enable the COT feature and make sure the COT datastore is configured as the order datastore. 
  This can be achieved programmatically (if necessary) with code like this:
   ```php
   add_filter(
       'woocommerce_order_data_store',
       function ( $default_data_store ) {
 	      return wc_get_container()->get( 'Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore' );
      },
       999,
       1
   );
   ```
2. Via custom code, try to create a new order with some properties/metadata/etc. For example:
   ```php
   $order = new WC_Order();
   $order->set_status( 'on-hold' );
   $order->set_created_via( 'php-code' );
   $order->set_currency( get_woocommerce_currency() );
   $order->set_customer_ip_address( WC_Geolocation::get_ip_address() );
   $order->set_customer_user_agent( wc_get_user_agent() );
   $order->set_billing_first_name( 'Albert' );
   $order->set_billing_last_name( 'Einstein' );
   $order->set_billing_email( 'albert.einstein@example.org' );
   $order->set_billing_address_1( 'Woo Street' );
   $order->set_billing_city( 'WooCity' );
   $order->set_billing_country( 'US' );
   $order->set_shipping_first_name( 'Elsa' );
   $order->set_shipping_last_name( 'Einstein' );
   $order->set_shipping_address_1( 'Woo Street' );
   $order->set_shipping_city( 'WooCity' );
   $order->set_shipping_country( 'US' );
   
   $order->get_data_store()->set_stock_reduced( $order, true, false );
   
   $payment_gateways = WC()->payment_gateways->payment_gateways();
   $order->set_payment_method( $payment_gateways['bacs'] );
   
   $order->set_shipping_total( 2.5 );
   $order->set_total( 12.5 );
   
   $order->save();
   ```
4. Confirm that tables `wc_orders`, `wc_order_operational_data`, `wc_order_addresses` and `wc_orders_meta` now have the data for the new order.

botwoo commented:
:bar_chart: Test reports for this pull request have been published and are accessible through the following links:

- [REST API Test Report](https://woocommerce.github.io/woocommerce-test-reports/pr/33341/api/)
- E2E Test Report (TBC)

Latest commit referenced in the reports: `First stab at OrdersTableDataStore::create()` d3254b239ade15cb3aa6af777c704a68828516b7
This comment will automatically be updated with the latest referenced commit when you push new changes to this pull request.

---

_Visit the [WooCommerce Test Reports](https://woocommerce.github.io/woocommerce-test-reports) homepage to view all published reports. See the [FAQs](https://github.com/woocommerce/woocommerce-test-reports/wiki/FAQs) page if you're having problems accessing them._

jorgeatorres committed:
`OrdersTableDataStoreHelper` no longer exists.

jorgeatorres committed:
First stab at `create()`

jorgeatorres committed:
Fix `order_stock_reduced` setters/getters

jorgeatorres committed:
Simplify handling of keys internal to the datastore

jorgeatorres committed:
Override `update_post_meta` to handle syncing of meta and props.

jorgeatorres committed:
Add changelog

jorgeatorres committed:
Add unit test for `create()`

jorgeatorres committed:
Rework `persist_order_to_db()` for simplicity

jorgeatorres committed:
Make PHPCS happy

jorgeatorres commented:
@vedanshujain: I'm assigning you as reviewer again because you reviewed the `update()` PR but let me know if you'd rather let other review. Thanks!

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php:
```diff
@@ -922,7 +956,36 @@ private function get_db_row_from_order_changes( $changes, $column_mapping ) {
 	 */
 	public function create( &$order ) {
-		throw new \Exception( 'Unimplemented' );
+		if ( '' === $order->get_order_key() ) {
```
>  we might need this for the COT datastore in order to reuse IDs from the wp_posts table, as long as that's necessary.

Let's use the ID generated by the posts table for ID in the orders table now, since it's one of our crucial safeguards it would be good to do it now.

jorgeatorres replied:
Done.

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php:
```diff
@@ -802,85 +823,98 @@ private function generate_select_clause_for_props( $table_alias, $props ) {
-				array_values( $update['format'] ),
-				$update['where_format']
-			);
+			$is_pk_table = ! empty( $update['pk_table'] );
```
as mentioned in another comment, it would be good to insert in posts table now and use that ID

jorgeatorres replied:
✔️

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php:
```diff
@@ -539,7 +539,28 @@ public function get_stock_reduced( $order ) {
 		// XXX implement $save = true.
 		$order = is_numeric( $order ) ? wc_get_order( $order ) : $order;
-		return $order->update_meta_data( '_order_stock_reduced', wc_string_to_bool( $set ) );
+		return $order->update_meta_data( '_order_stock_reduced', wc_bool_to_string( $set ) );
```
before launch, we should come back here and add logic to these small update in bulk, otherwise we are still doing lot of queries in inserting just one order.

jorgeatorres replied:
Yes. We definitely need to go through a bunch of TODO's and XXX's before launch. The blocker I added to the P2 post should remind us of these pending tasks, though I wouldn't mind creating a specific issue here in GH for that if you prefer that.

vedanshujain replied:
The blocker list is fine, I go through it from time to time, so we wouldn't miss it.

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php:
```diff
@@ -802,85 +823,98 @@ private function generate_select_clause_for_props( $table_alias, $props ) {
+	 * @return int      Order ID.
+	 * @throws \Exception If order data is not valid.
 	 */
 	protected function persist_order_to_db( $order, $only_changes = true ) {
```
what do you think about refactoring this method a bit so that:
1. Instead of manually building insert/update data for each table inline, we generate it via another method. This way, when other plugins also add their table in the future, they can use a filter in that method to add additional rows that need to be persisted? 
2. Either way, this method currently does a lot, it would be good to split this in smaller chunks so that we can test it bit easily.

jorgeatorres replied:
Thanks for the suggestion @vedanshujain! I ended up simplifying this method quite a bit (despite adding support for inserting records into the posts table).
Rows for COT tables are now generated in a different function, which also allows for some extensibility using a filter.

vedanshujain reviewed (commented):
Great work! I have added some questions that perhaps discussion although they are subjective.

jorgeatorres committed:
Do not override `post_type` when backfilling an order post

jorgeatorres committed:
Remove unnecessary reference to `$wpdb`

jorgeatorres committed:
Rework persisting logic

jorgeatorres committed:
Remove unnecessary check

jorgeatorres committed:
Revert "Do not override `post_type` when backfilling an order post"

This reverts commit efd1d4c48741e1efa6e69df1959c40de28bec923.

jorgeatorres commented:
Hey @vedanshujain!

Thanks for the initial review. I've addressed your feedback and simplified some of the methods.
Now the order ID is taken from the posts table, where we insert either a placeholder or a full order (that is later backfilled).

Please let me know what you think.

vedanshujain reviewed (commented):
LGTM, let's update the `since x.y.z` tags and we can approve and merge

jorgeatorres committed:
Docblock and method visibility fixes

jorgeatorres commented:
Hey @vedanshujain!

Thanks for taking a look 🥇 
I've updated the `@since` tag and also added a couple more. Now one of the checks is failing for... reasons (?) but all unit tests and code sniff check are passing.

Let me know if you feel some other change is required before merging. Thanks!

jorgeatorres committed:
Try to make CI happy (2.0)

psealock commented:
This should fix the highlight issue @jorgeatorres https://github.com/woocommerce/woocommerce/pull/33651

vedanshujain reviewed (approved):
LGTM! Although I see another failing test, feel free to merge if that's fine to ignore.

jorgeatorres commented:
Hey @vedanshujain! I'm merging this now. The check is still failing because [the fix](https://github.com/woocommerce/woocommerce/pull/33651) hasn't been merged yet, but apparently it has nothing to do with the code per se. 🤷

Here goes nothing!
